connector/connect/src/main/protobuf/spark/connect/relations.proto

```diff
@@ -457,3 +458,16 @@ message RenameColumnsByNameToNameMap {
   // duplicated B are not allowed.
   map<string, string> rename_columns_map = 2;
 }
+
+// Adding columns or replacing the existing columns that has the same names.
+message WithColumns {
+  // (Required) The input relation.
+  Relation input = 1;
+
+  // (Required)
+  //
+  // Given a column name, apply corresponding expression on the column. If column
+  // name exists in the input relation, then replacing the column. if column name
+  // does not exist in the input relation, then adding the column.
+  map<string, Expression> cols_map = 2;
```

Comment:
Can we use something here that guarantees an order? I don't think the current map based approach to withColumns was particularly well thought through (unordered maps will produce unstable schema's) and I don't necessarily want to tie myself to that.

Reply:
An alternative way of modelling this would be to use named expressions instead of key value pairs.

Reply:
This is an interesting topic. Given current withColumns API design, users cannot maintain or control the order over schema fields already (please correct me if I am wrong).

The nice thing to have is, if a user call withColumns by same parameter on the same DataFrame twice, the user see the same output schema through this proto (ordering not predictable but at least consistent)

However I don't know if this is feasible. For Python/Scala, we offer API like Dict/Map which once is used, the value iteration won't be deterministic so we cannot produce stable ordering on clients side already thus cannot preserve it through proto. Is this true? 

If we ever can stability produce an ordering on clients, we can of course maintain the ordering through proto. 

cc @cloud-fan @HyukjinKwon

Reply:
There is nothing we can do about the current APIs. I just think there is a very real chance that we will end up with a version of withColumns that takes an ordered collection of columns. A downside of the current APIs is that when you change platform versions you might end up with a different ordering.

One of the weirder things here will be is that the order you see on the client side (by iterating over the map), does not have to be the same order that is used by the server. This is because the map implementations used on both ends are different (e.g. use different hashes). This will be confusing to end users. I think that alone is a very strong reason to change this to a list of name-expression pairs.

Reply:
I see the reason to move to list of named expressions.

I also want to point out another thing so we have comprehensive discussion:

a list of named expressions means there could be duplicates while current API contract does not allow duplicates.  If an implementation submits a list of named expressions which contains duplicates, we gonna need to define a way to dedup. The best is if we can define a sorted unique map but maybe that is not easy to achieve by proto.

Reply:
IMO we just need to define what will happen to duplicates and document this. We can either throw an error, or pick one (proto3 would pick the last value).

Reply:
I can change this to named expression list. Will wait a bit to see if there is an opposite opinion.

Reply:
In fact, we can keep the order in `withColumns` by using `scala.collection.immutable.ListMap`. e.g.,) 

```scala
spark.range(1).withColumns(scala.collection.immutable.ListMap("id" -> $"id"))
```

So, it's up to users to decide if they care the order or not.

Reply:
Internally we convert them to a list of pairs so let's use list to keep the order from the map

Reply:
In addition, Python's dictionary keeps the insertion order by default.

Reply:
+1 on using repeated instead of map here to keep the order,

after the `Column` if refactored, will we use `Column` instead of `Expression` in proto messages?

Reply:
For proto I guess we will keep Expression. TBH I don't know why Spark chooses `Column` to refer to expressions....

Reply:
+1 on using a repeated tuple here. The reason is that the behavior should be defined by the API not by the language implementation. The reason we have consistent behavior in Spark today is that Python converts to Scala before even calling the method so there is only one client implementation.

My suggestion would be to change this to:

```
repeated Expression.Alias col_map
```

The Alias has a reference to an arbitrary expression and a name which is exactly what we want.